### PR TITLE
Signals and Slots rework for location service refactor

### PR DESCRIPTION
**Pull request by token419**
Refactored the location_service to use QWebsocket - removed global signals - added signals to existing apps

This makes the app shutdown immediately instead of being on a sleep loop if the sharing service thread was not started. This also removes the global signals that were in place for location sharing and moves them to a _signals dict off the main QApplication. I also moved the location sharing to a _services off the main QApplication as well so we can access them easier with QApplication.instance()

Pretty big pull, but check over the changes and let me know.

In location_service.py in the share_location function there is a commented out block for sending a fake location with a loc, i was using this in combination with commenting out the websocket.open() command to verify adding players/waypoints works, and on a live connection you can see them get cleaned up. Mostly useful for my testing while working on it, but we can remove it if we merge this in.

### CODE REVIEW

File: helpers/location_service.py [hunk 1]
```diff
@@ -1,166 +1,240 @@
-import json
-import ssl
-import threading
-import time
+from json import dumps, loads
```
Comment (rm-you) [inline helpers/location_service.py:1]: I'm not perfect about it, but best practice as I'm aware is to import modules, not classes/functions directly (though obviously QT has a different pattern, and they encourage class imports). Granted this is a style choice so I'm not going to harp on it too hard, but if possible would you mind leaving module imports where possible? Unless there's a counter-argument to this I haven't heard yet, which is also possible :P

File: helpers/location_service.py [hunk 2]
```diff
@@ -1,166 +1,240 @@
...
-    lsc = get_location_service_connection()
-    THREADPOOL.start(lsc)
+        # Set self.group_key - Handle settings changes and account for discord override
+        if config.data["sharing"]["discord_channel"]:
+            key = config.data["discord"]["url"].split("?")[0].split("/")[-1]
+            if key != "" and self.group_key != key:
+                self.group_key = key
+        else:
+            if self.group_key != config.data["sharing"]["group_key"]:
+                self.group_key = config.data["sharing"]["group_key"]
```
Comment (rm-you) [inline helpers/location_service.py:41]: Here and on 37 I'm not sure if there's a point in comparing the existing value with the new value, if they're equal then the assignment will be a no-op, just go ahead and set it IMO. Style choice I guess.

File: helpers/location_service.py [hunk 3]
```diff
@@ -1,166 +1,240 @@
...
+        self.host = config.data["sharing"]["url"]
+
+        # Set self.zone_name from config
+        self.zone_name = config.data["maps"]["last_zone"]
+
+        # Start the service from init if sharing is enabled in the config
+        if config.data.get("sharing", {}).get("enabled", False):
+            self.start()
+
+    # Used to start the websocket service and connect signals
```
Comment (rm-you) [inline helpers/location_service.py:53]: If you're making class/function level comments (I notice you do this for most of the functions in this file), just do them in docstring format! 👍

File: helpers/location_service.py [hunk 4]
```diff
@@ -1,166 +1,240 @@
...
+    def start(self):
+        # Setup Websocket
+        self.websocket = QWebSocket()
+        self.websocket.connected.connect(self.connected)
```
Comment (rm-you) [inline helpers/location_service.py:57]: lol it took me like three reads of this to figure out the "connect()" verb was for connecting a callback to a signal, not connecting a websocket, and "connected" was the event name XD

File: helpers/location_service.py [hunk 5]
```diff
@@ -1,166 +1,240 @@
...
+        self.websocket.error.connect(self.error)
+        self.websocket.disconnected.connect(self.disconnected)
+        self.websocket.textMessageReceived.connect(self.message)
+        self.websocket.open(QUrl(self.host))
```
Comment (rm-you) [inline helpers/location_service.py:61]: I need to test what happens if the user puts an empty-string in the hostname field (or in the sharing key) because users suck, though I guess worst case is they crash themselves, but then they'll go file an issue >_<

File: helpers/location_service.py [hunk 6]
```diff
@@ -1,166 +1,240 @@
...
+        QApplication.instance()._signals["maps"].location.disconnect(self.share_location)
+        QApplication.instance()._signals["maps"].death.disconnect(self.share_death)
+        QApplication.instance()._signals["locationsharing"].textMessageReceived.disconnect(self.parse)
+
+        # Set self.running to false
+        self.running = False
+
+    def config_updated(self):
+        # Handle performing character name override when saving a new name from settings
+        if config.data["sharing"]["enabled"] and config.data["sharing"]["player_name_override"] and config.data["sharing"]["player_name"] != self.character_name:
```
Comment (rm-you) [inline helpers/location_service.py:94]: I think if sharing is disabled we can quickly short circuit (and call the stop from line 110-112), but even if we keep going I think it's safe to update the sharing name regardless of whether it's enabled (and this is the only one you're bothering to check).

Long winded way of saying you could probably remove `config.data["sharing"]["enabled"] and` from the check above

File: helpers/location_service.py [hunk 7]
```diff
@@ -1,166 +1,240 @@
...
+    # Parses messages from the websocket server
+    def parse(self, websocket_message):
+        message = loads(websocket_message)
+        if message["type"] == "state":
+            # Only process locations if there is data for our location
+            if message.get("locations", False).get(self.zone_name.lower(), False):
+                # Interate through players in the zone
+                for player in message["locations"][self.zone_name.lower()]:
+                    #Process any players that are not us
+                    if player.lower() != self.character_name.lower():
```
Comment (rm-you) [inline helpers/location_service.py:148]: I think you also need to check `player.lower() != player_name_override`

File: helpers/location_service.py [hunk 8]
```diff
@@ -1,166 +1,240 @@
...
+                # Check players in the currently loaded QT map
+                for player in QApplication.instance()._parsers_dict["maps"]._map._data.players:
+                    # if the player is no longre in the zone and not ourselves, append them to be removed
+                    if player not in message["locations"][self.zone_name.lower()] and player != "__you__":
+                        players_to_remove.append(player)
+                for player in players_to_remove:
+                    QApplication.instance()._parsers_dict["maps"]._map.remove_player(player)
+
+            # Only process waypoints if there is data for our location
+            if message.get("waypoints", False).get(self.zone_name.lower(), False):
```
Comment (rm-you) [inline helpers/location_service.py:166]: first get() needs to return emptydict on lookup failure or the second get() will break

File: helpers/location_service.py [hunk 9]
```diff
@@ -1,166 +1,240 @@
...
+                # Iterate through the waypoints in the zone
+                for waypoint in message["waypoints"][self.zone_name.lower()]:
+                    w_data = message["waypoints"][self.zone_name.lower()][waypoint]
+                    w_point = MapPoint(x=int(w_data["x"]), y=int(w_data["y"]), z=int(w_data["z"]))
+                    w_icon = w_data.get("icon", "corpse")
```
Comment (rm-you) [inline helpers/location_service.py:171]: The server-side waypoints WILL have an icon field set for the single type we currently send (corpse), but there will probably be others... so maybe we default to something more generic than corpse? 😅

File: helpers/location_service.py [hunk 10]
```diff
@@ -1,166 +1,240 @@
...
+                    QApplication.instance()._parsers_dict["maps"]._map.add_waypoint(waypoint, w_point, w_icon)
+
+            # Remove waypoints that aren't in the zone
```
Comment (rm-you) [inline helpers/location_service.py:174]: I don't know if we can safely remove every waypoint? People can still have "local" waypoints, would this be removing them? Augh, it's been too long for me to remember some of the specifics of how this worked, will need to look again / test.

**Review by rm-you — CHANGES_REQUESTED** [review]
I think if you can fix a couple quick things in here, we can merge this and I can do some more extensive testing on it and we can fix any minor bugs that pop up in a future PR.

File: parsers/maps/window.py [hunk 11]
```diff
@@ -63,106 +71,42 @@ def __init__(self):
...
+            QApplication.instance()._signals["maps"].new_zone.emit(text[17:-1])
             self._map.load_map(text[17:-1])
         elif ZONE_MATCHER.match(text):
             new_zone = ZONE_MATCHER.match(text).groupdict()['zone'].lower()
             new_zone = MapData.translate_who_zone(new_zone)
             if new_zone not in (self._map._data.zone.lower(), 'everquest'):
+                QApplication.instance()._signals["maps"].new_zone.emit(new_zone)
                 self._map.load_map(new_zone, keep_loc=True)
         elif text[:16] == 'Your Location is':
+            QApplication.instance()._signals["maps"].location.emit(timestamp.isoformat(), text[17:])
```
[the remark this answers is not in the document]
Comment (token419) [reply parsers/maps/window.py:88]: I just matched what was already being done, no preference on timezone format here.

File: parsers/maps/window.py [hunk 12]
```diff
@@ -63,106 +71,42 @@ def __init__(self):
...
     def parse(self, timestamp, text):
         if text[:23] == 'LOADING, PLEASE WAIT...':
-            pass
+            QApplication.instance()._signals["maps"].zoning.emit()
         elif text[:16] == 'You have entered':
+            QApplication.instance()._signals["maps"].new_zone.emit(text[17:-1])
             self._map.load_map(text[17:-1])
         elif ZONE_MATCHER.match(text):
             new_zone = ZONE_MATCHER.match(text).groupdict()['zone'].lower()
             new_zone = MapData.translate_who_zone(new_zone)
```
[the remark this answers is not in the document]
Comment (rm-you) [reply parsers/maps/window.py:82]: I think that IS the key we end up translating to, but there has to be a mapping between the text name of the zone that's printed in logs on zone in (and on /who which is surprisingly sometimes different -- see Kael) and what we show the user in the manual map selection dropdown... just wish it was a like bit cleaner, but I don't know if we can do much about it.

File: helpers/location_service.py [hunk 13]
```diff
@@ -1,166 +1,239 @@
...
+    # Websocket message handler - handles any incoming messages from the websocket server
+    def websocket_message(self, message):
+        QApplication.instance()._signals["locationsharing"].textMessageReceived.emit(message)
+
+    # Parses messages from the websocket server
+    def parse(self, websocket_message):
+        message = json.loads(websocket_message)
+        if message["type"] == "state":
+            # Only process locations if there is data for our location
+            if message.get("locations", False).get(self.zone_name.lower(), False):
```
Comment (rm-you) [inline helpers/location_service.py:143]: Oh, I missed one of these -- needs to be emptydict on the first get() here too

**Review by rm-you — APPROVED** [review]
